Improve compilation and linking of Rust files

wmitros opened the pull request:
This patch can be treated as a follow-up to #10341, or as a preparation before migrating
from the lacking C++ wasmtime bindings to our custom implementation that will use the
original Rust crate.

The main changes are the introduction of cargo profiles (equivalents of scylla build modes), fixes related to linking against multiple rust packages and corresponding adjustments in documentation.

Some details are included in the commit messages, but you can read more about using 
the Rust "CXX" bridge in a foreign build system [here](https://cxx.rs/build/other.html) and the available cargo profile options [here](https://doc.rust-lang.org/cargo/reference/profiles.html)

wmitros commented:
I will be away next week, but feel free to comment anyway

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1470/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1472/

wmitros commented on configure.py:
```diff
@@ -1406,7 +1402,7 @@ def clang_inline_threshold():
         modes[mode]['cxxflags'] += f' -Wstack-usage={modes[mode]["stack-usage-threshold"]} -Wno-error=stack-usage='
 
-has_wasmtime = os.path.isfile('/usr/lib64/libwasmtime.a') and os.path.isdir('/usr/local/include/wasmtime')
+has_wasmtime = False
```
I added this because linking with both `libwasmtime` and `librust_combined` was causing duplicate symbol errors. This will not be an issue anymore when we add wasmtime in rust, and that patch is only waiting for #10306, so this temporary fix should not last a long time (wasm is also still experimental)

nyh replied:
But this "fix" means that wasm will no longer be working at all, until a later fix? @psarna are we fine with that?
I see 10306 has already been merged, so can't you fix that wasm issue right now and not have any temporary period where it doesn't work?

psarna replied:
Oh, I missed that one. No, it's not acceptable, we have experimental support, but we also already published a few examples of it online and we *expect* people to try it out, not to find out they're missing in a particular commit. What should be done here is not linking with librust_combined now if it's useless and doesn't provide anything. Or perhaps it's enough to mark it no-std? ref: https://docs.rust-embedded.org/book/intro/no-std.html

wmitros replied:
After trying a couple of methods, it seems that to avoid the linking issues we would have to at least modify our dependencies (in combination with using no-std).
We can't avoid linking with librust_combined as long as we have the `test/boost/rust_test`.
We can't merge the mentioned rust wasmtime patch before this one, because similarly, we'll have linking issues.
Having that in mind, I think it would be best if we combined these 2 patches in one PR. I'll soon publish the new PR depending on this one anyway, but recently we decided that there's something I still need to add to it, so you'll have to wait a bit.

psarna replied:
I'm also fine with temporarily dropping rust_test if it helps, we can bring it back once we have wasmtime, or just drop it in favor of some custom wasmtime tests

wmitros replied:
I see you commented this before yesterday's patch. Would you still like rust_test dropped or will it be fine if we just combine this series with the wasmtime one and try to merge them at the same time?

psarna replied:
if dropping the test means that we can cheaply avoid disabling wasm-based udfs for a while (even within the same series), then I think it's worth it. But if it's in any way not trivial, the current state is acceptable too, because wasm support is experimental anyway.

piodul commented on rust/Cargo.toml:
```diff
@@ -9,3 +9,31 @@ inc = { path = "inc", version = "0.1.0" }
+
+[profile.rust-sanitize]
+inherits = "dev"
+opt-level = 2
```
The `sanitize` mode compiles cpp files using `-Os`, so maybe 's' would be more appropriate here?

wmitros replied:
I expect that both clang and rustc use similar optimization passes for the `s` optimization level considering they both use llvm, but I didn't find any direct comparison that confirms it. There is a slight difference in how they are described in their official documentations though: clang describes `-Os` as "Like `-O2` with extra optimizations to reduce code size", while rust describes it as just: "optimize for binary size".
The description of the `Sanitize` mode provided by scylla is "a mode with optimizations and sanitizers enabled, used for finding memory errors", so I used the mode that is confirmed to at least have some optimizations.

On the other hand, I haven't confirmed that the `-O2` mode in rust is that similar to clang's `-O2`, so for the sake of consistency, maybe the `s` option would be best after all

What do you think?

piodul replied:
I think that the main purpose of `-Os` is to "optimize for size", so IMO we should use `s` here.

wmitros replied:
OK

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1627/

wmitros commented on .gitignore:
```diff
@@ -31,3 +31,4 @@ docs/poetry.lock
 compile_commands.json
 .ccls-cache/
 .mypy_cache
+rust/Cargo.lock
```
I did base my decision to put the Cargo.lock here on the same article, but I see that your interpretation makes more sense.
Tracking the Cargo.lock will however require an additional step of regenerating it after any dependency changes, so I'll mention it in the docs

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1649/

avikivity commented:
I don't understand any of this. We have a build running in install-dependencies.sh, why do we need to run it again?

wmitros commented:
> I don't understand any of this. We have a build running in install-dependencies.sh, why do we need to run it again?

Maybe the naming here is misleading. This patch affects the compilation of source files in the `rust` directory, while install-dependencies.sh is only used for downloading the rust toolchain needed for the compilation (`cargo` and a `cxxbridge` script for generating C++ headers from rust files)

wmitros commented:
I noticed cxxbridge allows to remove some code duplication using `cxx.h` instead of `cxx.hh` so I added this in the rebase

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1661/

avikivity commented:
> > I don't understand any of this. We have a build running in install-dependencies.sh, why do we need to run it again?
> 
> Maybe the naming here is misleading. This patch affects the compilation of source files in the `rust` directory, while install-dependencies.sh is only used for downloading the rust toolchain needed for the compilation (`cargo` and a `cxxbridge` script for generating C++ headers from rust files)

Please try to clarify the wording then so I can understand it.

nyh commented on rust/Cargo.toml:
```diff
@@ -9,3 +9,30 @@ inc = { path = "inc", version = "0.1.0" }
+
+[profile.rust-sanitize]
+inherits = "dev"
+opt-level = "s"
```
What does "opt-level=s" do?

psarna replied:
binary size: https://doc.rust-lang.org/cargo/reference/profiles.html

nyh replied:
Why does "sanitize" mode specifically needs to optimize for size? I'm not saying it's bad, it just sounds strange.

piodul replied:
The cpp files are compiled with the `-Os` flag in the sanitize mode, so I suggested using an equivalent flag for rust files.

As for why we are optimizing for size in the cpp files - I'm not sure. The original patchset which introduces sanitize mode (https://groups.google.com/g/scylladb-dev/c/wQ_pw5zkJIg/m/bHW-oeO3AwAJ) doesn't seem to explain it, and the author doesn't work with us anymore. If you have a good argument not to optimize for size here, then I'm completely fine with using a different mode.

nyh replied:
No, I'm fine with this optimization mode. Any other choice will seem equality arbitrary...

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1737/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1874/

wmitros commented:
In the last rebases I modified the configure script: because we're creating one library for all rust sources together, I added a list of all rust files that are required for the lib. The targets that require rust sources should now include the related `lib.rs` file 
I've also experimented with auto-generating the `Cargo.lock` file, but once we track it and it's missing, we should probably fail and require that it's pulled instead of generating a new one

wmitros commented:
This patch is now also a part of #11351 so if it makes more sense to comment with the context of further changes, it can be done there

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1886/

avikivity commented:
Looks reasonable, will defer to @psarna decision.

psarna commented:
Since it's also part of https://github.com/scylladb/scylladb/pull/11351, I think we should close this issue and continue with https://github.com/scylladb/scylladb/pull/11351. The period in which wasm-based UDFs are disabled should be minimized/removed, and the most sure way of that is including these changes along with https://github.com/scylladb/scylladb/pull/11351

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1982/

wmitros commented:
This PR was added as a part of #11351, which just got merged. We can close this now as well
